### ISSUE TITLE
Automatic plugin enabling fails while in a monorepo

### ISSUE THREAD

**Issue by Zamiell** [webpro/knip#176]
Hello and thanks for the awesome tool!

Currently, `knip` appears to work by examining the closest package.json file for an `eslint` dependency. However, this algorithm will fail in a monorepo that looks like this:

```
my-monorepo/
├── package.json (contains development deps for every subpackage)
└── packages/
    └── foo
        └── package.json (contains only production deps, for the purposes of publishing to npm)
```

AFAIK, it is common in monorepos to not repeat the ESLint dep for N subpackages. Instead, the ESLint dep is declared once in the root package.json file.

Currently, it is possible to work around this problem by creating a `knip.json` file for every subpackage in the repo, containing `ignore: [".eslintrc.cjs"]`. However, this is of course quite cumbersome. It would be nice to update the algorithm such that it can properly detect that ESLint is a dep while inside of a monorepo.

Finally, I'll note that the same problem applies to the Jest plugin, the Remark plugin, and so on.

**Comment by webpro** [comment]
Any chance you have a rep(r)o I can look at?

**Comment by Zamiell** [comment]
Yeah:

- `git clone git@github.com:IsaacScript/isaacscript.git`
- `cd isaacscript`
- `yarn`
- `cd packages/eslint-plugin-isaacscript`
- `npx webpro` (notice that it succeeds)
- `vim knip.json` (delete the exception for ".eslintrc.json")
- `npx webpro` (notice that it fails, even though "eslint" is a dep in the root package.json of the monorepo)

**Comment by webpro** [comment]
Thanks for the repo. A few notes about how Knip currently works:

- Knip support for monorepos means that there's `workspaces` (e.g. `["packages/*"]`) in a root `package.json` or `knip.json` and that Knip is ran from the root. Then individual packages will not report unused dependencies if the root has that dependency. Knip does not support reading them from Nx `workspace.json`.
- Knip doesn't look in `peerDependencies` and it doesn't look in parent directories relative to where you run it from (it's scoped to the directory you run it from).
- Each package having an `.eslintrc.js` without the referenced eslint dependencies in either the package itself or the root package means you would have to ignore that file indeed. Creating a root `knip.json` with `workspaces` allows to configure things in a single file.

**Comment by Zamiell** [comment]
Thanks for the quick reply.

> it doesn't look in parent directories relative to where you run it from (it's scoped to the directory you run it from).

In the future, this would be a great feature for `knip` to have! Especially considering that this is how ESLint works (and Prettier, and CSpell, and so on), e.g. looking backward until it finds a matching config.

In a hypothetical where this feature already exists, then the work required to solve this specific monorepo issue would be mostly completed.

**Comment by webpro** [comment]
Knip might feel similar to ESLint, but Knip is also quite different in scope. You would also need to compare it with TypeScript and Nx with their multiple projects/packages and dependency graphs, etc.

Compared to your "looking backward" it basically works the other way around. You should run it from the root, and from there you can analyze individual packages using the `--workspace` argument.

**Comment by Zamiell** [comment]
> You should run it from the root, and from there you can analyze individual packages using the --workspace argument.

This doesn't work in my monorepo. I looked at the README file, and it says that the --workspace flag is only supported if you have one of these three things:

> In the workspaces array in package.json (npm, Yarn, Lerna).
In the workspaces.packages array in package.json (legacy).
In the packages array in pnpm-workspace.yaml (pnpm).

Nx doesn't use the workspaces feature of the package manager; it is unnecessary. What is the recommended way to use `knip` in an Nx monorepo?

**Comment by webpro** [comment]
How about this:

```json
{
  "$schema": "https://unpkg.com/knip@2/schema.json",
  "ignoreWorkspaces": ["packages/isaacscript-lua"],
  "workspaces": {
    "packages/*": {}
  }
}
```

**Comment by Zamiell** [comment]
Thanks. I tried out that config, but when executing `npx knip --workspace=eslint-plugin-isaacscript` from the root of the project, it still gives the same behavior as before: `knip` just automatically succeeds without any particular error message, skipping all of the issues.

As an aside, we should update the README to include the undocumented "workspaces" field as the 4th option that it searches for.

**Comment by webpro** [comment]
> Thanks. I tried out that config, but when executing `npx knip --workspace=eslint-plugin-isaacscript` from the root of the project, it still gives the same behavior as before: `knip` just automatically succeeds without any particular error message, skipping all of the issues.

You would need to use the workspace folder (e.g. `--workspace packages/folder`), not the package name.

Would be good to raise an error when the folder doesn't exist.

> As an aside, we should update the README to include the undocumented "workspaces" field as the 4th option that it searches for.

Agreed, will do!

**Comment by Zamiell** [comment]
> You would need to use the workspace folder (e.g. --workspace packages/folder), not the package name.

Thanks, that worked!

However, now I'm experiencing another problem: `knip` is not reading the `knip.json` file in the workspace directory. Is that intended? As a workaround, I could consolidate the individual package `knip.json` files into the root `knip.json`, but that feels wrong, as project-specific configuration belongs in the project-specific directories.

> Would be good to raise an error when the folder doesn't exist.

Yes please. : )

**Comment by webpro** [comment]
> However, now I am experiencing another problem: `knip` is not reading the `knip.json` file in the workspace directory. Is that intended? As a workaround, I could consolidate the individual package `knip.json` files into the root `knip.json`, but that feels wrong, as project-specific configuration belongs in the project-specific directories.

That's intended.

You can handle each package separately, but as discussed Knip is then scoped to that package only.

If you want more flexibility, you could consider using `knip.ts` for dynamic configurations to make it more DRY.

**Comment by Zamiell** [comment]
> If you want more flexibility, you could consider using knip.ts for dynamic configurations to make it more DRY.

Ah, that's a good solution, thanks - I'll have the root `knip.ts` import from `packages/foo/knip.ts`.

**Comment by webpro** [comment]
> that feels wrong, as project-specific configuration belongs in the project-specific directories.

By the way, in general I agree with this sentiment, this is also why Knip features a `--production` and a `--strict` flag. Yet it falls apart when we start looking up configuration files in parent directories or dependencies from parent packages. Eventually Knip is full of assumptions and heuristics in a highly dynamic ecosystem.

Will keep this ticket open to track:

- Improve documentation re. workspace config
- Throw when `--workspace [dir]` doesn't exist

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v2.19.0. See [Release 2.19.0](https://github.com/webpro/knip/releases/tag/2.19.0) for release notes._

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v2.19.0. See [Release 2.19.0](https://github.com/webpro/knip/releases/tag/2.19.0) for release notes._

**Comment by Zamiell** [comment]
Looks like the bot is bugger - it sent a duplicated message.

**Comment by webpro** [comment]
Yes, ideally it shouldn't send it twice, but I mentioned `resolves #176` in two different commits (for two separate issues).
